src/libraries/System.Text.RegularExpressions/tests/UnitTests/SymbolicRegexTests.cs

```diff
@@ -0,0 +1,149 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Globalization;
+using Xunit;
+using System.Text.RegularExpressions.Symbolic;
+using System.Collections.Generic;
+
+namespace System.Text.RegularExpressions.Tests
+{
+    public class SymbolicRegexTests
+    {
+        [Theory]
+        [InlineData(UnicodeCategory.DecimalDigitNumber, 370)] //37 different kinds of decimal digits
+        [InlineData(UnicodeCategory.Surrogate, 2048)]         //1024 low surrogates and 1024 high surrogates
+        public void BDDCardinalityTests(UnicodeCategory category, uint expectedCardinality)
+        {
+            BDD digits = UnicodeCategoryConditions.GetCategory(category);
+            Assert.Equal(expectedCardinality, ComputeCardinality(digits));
+
+            // Returns how many characters this BDD represents
+            static uint ComputeCardinality(BDD bdd)
+            {
+                if (bdd.IsEmpty)
+                {
+                    return 0;
+                }
+                if (bdd.IsFull)
+                {
+                    return ushort.MaxValue;
+                }
+
+                (uint Lower, uint Upper)[] ranges = BDDRangeConverter.ToRanges(bdd);
+                uint result = 0;
+                for (int i = 0; i < ranges.Length; i++)
+                {
+                    result += ranges[i].Upper - ranges[i].Lower + 1;
+                }
+                return result;
+            }
+        }
+
+        public static IEnumerable<object[]> SafeThresholdTests_MemberData()
+        {
+            var charSetSolver = new CharSetSolver();
+            var bddBuilder = new SymbolicRegexBuilder<BDD>(charSetSolver, charSetSolver);
+            var converter = new RegexNodeConverter(bddBuilder, null);
+
+            // all patterns are considered with RegexOptions.ExplicitCapture
```

Comment:
Should we have tests for captures as well / validate that they do / don't impact the NFA size as expected?

Reply:
Yes, more tests are needed for that.

Reply:
In fact there is a whole set of tests missing that reflect the actual NFA construction and then comparison with the estimated size.